metagraph/src/graph/annotated_dbg.cpp

```diff
@@ -63,7 +65,12 @@ ::annotate_sequence(std::string_view sequence,
             return;
         }
     }
-
+    if (taxonomy != nullptr) {
+        uint64_t lca = 0;
+        if (taxonomy->find_lca(labels, lca) ) {
+            taxonomy->update_taxonomic_map(indices, lca);
+        }
+    }
```

Comment:
I think that it would be faster to wait until the metagraph annotation is finished and then call this `update_taxonomic_map` only once per annotation row. This approach would allow multi threading.

Reply:
update: parsing all the kmers from annotation matrix will be certainly slower, because, currently it is calling LCA function for entire unitigs

Reply:
Why do you pass `indices` to `taxonomy->update_taxonomic_map`? How are they used?
Isn't the decision only based on the label/tax_id counts?

Reply:
`indices` is a list of kmers that exists in one virus sequence. This sequence has an accession version as label (vector `labels` which has always size 1 in the virus sequences used for testing). The lookup table is used for translating accession version to taxid.

`taxonomy->taxonomic_map[kmer]` is returning the LCA taxid for the give kmer. As we can have this kmer in different viruses, we are sending it inside `indices` param of `taxonomy->update_taxonomic_map` multiple times and constantly calculate the new_lca considering the old value stored: `taxonomy->taxonomic_map[kmer] =lca(taxonomy->taxonomic_map[kmer] , lca)`. 

This `taxonomy->taxonomic_map[kmer]` will be exported by TaxoDB https://github.com/ratschlab/metagraph/pull/283/files#diff-0913e659af3790404c8335af51b605365a2bd1400acb8abd9b20b99cc66b2df8R147-R153

imported by TaxoClassifier constructor https://github.com/ratschlab/metagraph/pull/285/files#diff-ec92d55b97b8ad3792ad7db9bebbeed43041c069c1655338ca7588ddce7c7ce3R37-R41

and will be further used by TaxoClassifier assign_class method https://github.com/ratschlab/metagraph/pull/285/files#diff-ec92d55b97b8ad3792ad7db9bebbeed43041c069c1655338ca7588ddce7c7ce3R51-R61

Reply:
The k-mers are associated with accession ids, and these accession ids are associated with taxonomy ids, right?

When you map k-mers to tax ids, you can first map them to accession ids (labels) and then map these labels to taxonomy ids, using the mapping.
The second mapping can be done without knowing the k-mer identities, thus, there is no need to pass the list `indices`.

Did I miss something?

Reply:
Never mind, after looking at https://github.com/ratschlab/metagraph/pull/283/files#diff-1094a6617a94f28060d4bc862d4798543266da5cf04099258d2cdf512cc23d30R254-R265, I see what is going on.

Back to your original comment.
I agree, doing this after the annotation is constructed should be faster.
Note that currently, you're computing LCA ]for each k-mer in a unitig, and not for entire unitigs.

Reply:
Indeed it is doing LCA per k-mer, I think that I was tired last evening and forgot that there are 2 round of doing LCA. A change that is doing the taxoDB construction after the annotation is on this branch: https://github.com/ratschlab/metagraph/compare/taxonomic_anno...annotation_to_taxo_map

I tested both of them on the small dataset (first 10^5 lines of the virus sequences) and the current version is going through `taxonomy->update_taxonomic_map` almost instantly (1-2 seconds), while for parsing the annotation matrix goes for 74.88 seconds. I created a new dbg graph for those 10^5 lines and not used the big dbg one.

Is there any other faster way to call the kmers excepting the one that I tried? https://github.com/ratschlab/metagraph/compare/taxonomic_anno...annotation_to_taxo_map#diff-c861be87c8e771e7b0b11ba640aa44846ebcb312c32593322ad2d3127f5ff40eR287-R296

Otherwise it could be that `normalized_taxid[lookup_table[accession_version]]` is very slow and avoiding a call of those 2 maps for each kmer comes with a huge time improvement

Reply:
That other branch is doing it per k-mer as well.
Which is fine. But I think we need to do it differently.

Currently, you're storing LCA in a hash table. Imagine how it's going to grow for graph sizes about 1 trillion.

We'd need to make two changes.
1. First, change the format of the tax-id annotation, to make it space-efficient, but still fast to query.
As far as I understand, we need to store a single LCA tax-id for each k-mer.
This means, storing an array of integers. We could just store them in sdsl::int_vector<> for the first version, and implement compression later, maybe similarly to the vector of row-codes in rainbowifsh
https://github.com/ratschlab/metagraph/blob/270b38ad6ff1ceee441d6be32d6433450d7808bc/metagraph/src/annotation/binary_matrix/rainbowfish/rainbow.hpp#L54:L55.

2. For construction, the fastest way would be to do it by iterating columns.
So, first, we construct the full annotation matrix (the labels are accession IDs).
Then, iterate through columns and update the LCA for all bits set in that column.
To do it all faster and with limited RAM, we would process columns in batches and keep the intermediate list of computed LCA in vector streamed from disk.

Reply:
nice, thank you! Will take a look and implement these 2 ideas